src/cache.rs

```diff
@@ -0,0 +1,64 @@
+use std::{collections::HashMap, path::PathBuf};
+
+use crate::Tileset;
+
+pub type ResourcePath = PathBuf;
+
+pub trait ResourceCache {
+    fn get_tileset(&self, path: &ResourcePath) -> Option<&Tileset>;
+    fn get_or_insert_tileset(&mut self, path: ResourcePath, tileset: Tileset) -> &Tileset;
+    fn get_or_insert_tileset_with<F>(&mut self, path: ResourcePath, f: F) -> &Tileset
+    where
+        F: FnOnce() -> Tileset;
+    fn get_or_try_insert_tileset_with<F, E>(
+        &mut self,
+        path: ResourcePath,
+        f: F,
+    ) -> Result<&Tileset, E>
+    where
+        F: FnOnce() -> Result<Tileset, E>;
+}
+
+pub struct DefaultResourceCache {
+    tilesets: HashMap<ResourcePath, Tileset>,
+}
+
+impl DefaultResourceCache {
+    pub fn new() -> Self {
+        Self {
+            tilesets: HashMap::new(),
+        }
+    }
+}
+
+impl ResourceCache for DefaultResourceCache {
+    fn get_tileset(&self, path: &ResourcePath) -> Option<&Tileset> {
+        self.tilesets.get(path)
+    }
+
+    fn get_or_insert_tileset(&mut self, path: ResourcePath, tileset: Tileset) -> &Tileset {
+        self.tilesets.entry(path.to_owned()).or_insert(tileset)
+    }
+
+    fn get_or_insert_tileset_with<F>(&mut self, path: ResourcePath, f: F) -> &Tileset
+    where
+        F: FnOnce() -> Tileset,
+    {
+        self.tilesets.entry(path.to_owned()).or_insert_with(f)
+    }
+
+    fn get_or_try_insert_tileset_with<F, E>(
+        &mut self,
+        path: ResourcePath,
+        f: F,
+    ) -> Result<&Tileset, E>
+    where
+        F: FnOnce() -> Result<Tileset, E>,
+    {
+        if !self.tilesets.contains_key(&path) {
+            Ok(self.tilesets.entry(path).or_insert(f()?))
+        } else {
+            Ok(self.tilesets.get(&path).unwrap())
+        }
```

Comment:
Is there a better way to do this?

Reply:
I guess since loading the tileset might fail we can't use `or_insert_with`. I've tried:
```rust
        Ok(self.tilesets.entry(path).or_insert_with(|| {
            f()?
        }))
```
But it results in something I don't understand:
```
error[E0277]: the trait bound `tileset::Tileset: FromResidual<Result<Infallible, E>>` is not satisfied
   --> src/cache.rs:63:53
    |
63  |           Ok(self.tilesets.entry(path).or_insert_with(|| {
    |  ______________________________________--------------_^
    | |                                      |
    | |                                      required by a bound introduced by this call
64  | |             f()?
65  | |         }))
    | |_________^ the trait `FromResidual<Result<Infallible, E>>` is not implemented for `tileset::Tileset`
    |
note: required by `from_residual`
   --> /home/bjorn/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/try_trait.rs:339:5
    |
339 |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Reading [`FromResidual`](https://doc.rust-lang.org/std/ops/trait.FromResidual.html) hasn't really helped me, but I doubt it's going to achieve what we want, namely that a potential error in the callback is returned by `or_insert_with`.

But, we could at least avoid the double hash map lookup:

```suggestion
        Ok(match self.tilesets.entry(path) {
            Entry::Occupied(entry) => entry.into_mut(),
            Entry::Vacant(entry) => entry.insert(f()?),
        })
```

Reply:
> But, we could at least avoid the double hash map lookup

Ah, that's it. Didn't know Entry was an enum.

Reply:
> You can't use `?` because the parent closure should return `T`, not `Result<T, E>`.

Right, and apparently for some types one could implement `FromResidual<Result<Infallible, E>>` so that you get a valid object even in the case of error.

> Ah, that's it. Didn't know Entry was an enum.

Right, in fact the `Entry` seems to be very much like your `LayerTile`, in that it enables getting references to elements with the lifetime of their container.